File: src/lista_encadeada.cpp
#include "lista_encadeada.h"

void lista_encadeada::buscaElemento(int elemento){
	std::cout << "Busca: tentativa de bloqueio do mutex.\n";
	mtx.lock();
	std::cout << "Busca: mutex bloqueado.\n";

	bool achou = false;

	for(auto it{this->lista.begin()}; it != this->lista.end(); ++it){
		if(*it == elemento){
			achou = true;
		}
	}

	if(achou){
		std::cout << "\tBusca: elemento " << elemento << " encontrado.\n";
	}else{
		std::cout << "\tBusca: elemento " << elemento << " NÃO encontrado.\n";
	}

	std::cout << "Busca: mutex liberado.\n";
	mtx.unlock();
}

void lista_encadeada::removeElemento(int elemento){
	std::cout << "Remoção: tentativa de bloqueio do mutex.\n";
	mtx.lock();
	std::cout << "Remoção: mutex bloqueado.\n";

	this->lista.remove(elemento);
	std::cout << "\tRemoção: elemento " << elemento << " removido (se existente).\n";

	std::cout << "Remoção: mutex liberado.\n";
	mtx.unlock();
}

void lista_encadeada::insereElemento(int elemento){
	std::cout << "Inserção: tentativa de bloqueio do mutex.\n";
	mtx.lock();
	std::cout << "Inserção: mutex bloqueado.\n";
	
	auto it = this->lista.before_begin();
	auto it2 = it;
	while(it != this->lista.end()){
		++it;
		if(it != this->lista.end()){
			it2 = it;
		}
	}
	this->lista.insert_after(it2, elemento);
	std::cout << "\tInserção: elemento " << elemento << " inserido.\n";
	
	std::cout << "Inserção: mutex liberado.\n";
	mtx.unlock();
}

File: ListaEncadeadaConcorrente/src/ListaEncadeada.java
import java.util.LinkedList;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class ListaEncadeada {

    private LinkedList<Integer> lista;
    private ReentrantReadWriteLock lock;
    private Lock leitura;
    private Lock escrita;

    public ListaEncadeada(){
        lista = new LinkedList<>();
        lock = new ReentrantReadWriteLock(true);
        leitura = lock.readLock();
        escrita = lock.writeLock();
    }

    public void buscaElemento(int elemento, String nomeThread){
        System.out.println(nomeThread + ": tentou bloquear o readLock.");
        leitura.lock();
        System.out.println(nomeThread + ": bloqueou o readLock.");
        try {
            int posicao = lista.indexOf(elemento);
            if(posicao != -1){
                System.out.println("\t" + nomeThread + ": elemento " + elemento + " encontrado no posição " + posicao + ".");
            }else{
                System.out.println("\t" + nomeThread + ": elemento " + elemento + " não encontrado.");
            }
        } catch (Exception e){
            e.printStackTrace();
        } finally {
            System.out.println(nomeThread + ": liberou o readLock.");
            leitura.unlock();
        }
    }

    public void insereElemento(int elemento, String nomeThread){
        System.out.println(nomeThread + ": tentou bloquear o writeLock.");
        escrita.lock();
        System.out.println(nomeThread + ": bloqueou o writeLock.");
        try {
            lista.add(elemento);
            System.out.println("\t" + nomeThread + ": elemento " + elemento + " inserido.");
        } catch (Exception e){
            e.printStackTrace();
        } finally {
            System.out.println(nomeThread + ": liberou o writeLock.");
            escrita.unlock();
        }
    }

    public void removeElemento(int elemento, String nomeThread){
        System.out.println(nomeThread + ": tentou bloquear o writeLock.");
        escrita.lock();
        System.out.println(nomeThread + ": bloqueou o writeLock.");
        try {
            boolean sucesso = lista.remove((Integer) elemento);
            if(sucesso)
                System.out.println("\t" + nomeThread + ": elemento " + elemento + " removido.");
            else{
                System.out.println("\t" + nomeThread + ": elemento " + elemento + " inexistente, remoção não realizada.");
            }
        } catch (Exception e){
            e.printStackTrace();
        } finally {
            System.out.println(nomeThread + ": liberou o writeLock.");
            escrita.unlock();
        }
    }

}